### ISSUE TITLE
Relax validation of hub.db.type

### ISSUE THREAD

**Issue by weisdd** [jupyterhub/zero-to-jupyterhub-k8s#2246]
### Proposed change

With the newly added schema validation (thanks for that, btw!), it's no longer possible to specify a custom database type (outside of the predefined set). It might be a good idea to relax schema validation for `hub.db.type`.

In our environment, we decided to stick with CocroachDB instead of PostgreSQL due to easier clusterization and scalability.
To make it work, we had to do the following:
* install sqlalchemy-cocroachdb before hub starts;
* fix a tiny defect in jupyterhub (https://github.com/jupyterhub/jupyterhub/pull/3484, it was not part of `1.4.1`).
So far, everything seems to work well.

The real question here is the following:
Should `hub.db.type` represent only the set of database drivers that are available out-of-the-box?

### Alternative options

As far as I know, in our case, specifying `postgres` is not going to have any negative effects. It only instructs `jupyterhub_config.py` to set an additional environment variable. So, we can live with that.

If you think that `hub.db.type` should stay as it is now, maybe adding a special comment like ("set it to `postgres` in case you decide to rely on an alternative sqlalchemy driver").

Personally, I'm fine with any of the options. Just having a clarification on what's the recommended approach would be great as well.

### Who would use this feature?

I saw somewhere in jupyterhub's docs that it should work with any sqlalchemy driver. So, I guess, I'm not the only one who relies on cocroachdb on something else.

**Comment by consideRatio** [comment]
I'm somewhat open to supporting other DB providers by installing some python package as well, and setting some password in an env var like we do for `hub.db.url` as well if someone champions it. I just don't want there to be for example 10+ additional providers added.

**Comment by weisdd** [comment]
@consideRatio If we expand a number of supported DB providers, that will set an implicit expectation that jupyterhub is thoroughly tested with all of them, which is not the case (I guess it requires adjustments to tests in CI).
On the other hand, if we just add the "other" option to `values.schema.json`, it'll sound like "use it at your own risk", which is a fair statement.

So, I think, `other` is a preferable option for now.
And if, at a certain point, we feel comfortable expanding the list of officially supported DB providers, then great, we can do that as well :)

**Comment by weisdd** [comment]
@consideRatio As for installing packages for additional DB drivers:

Currently, I install it like this:

```yaml
hub:
  command:
    - "/usr/bin/bash"
    - "-ec"
    - set -exo pipefail; pip3 install -Iv sqlalchemy-cockroachdb==1.3.3; exec jupyterhub "$@"
```

Of course, it'd be great to have it pre-installed. Would it be fine if I open a PR with exactly that change (just adding the driver to the requirements) or there's a bigger list of things to do before I can expect such a change to be potentially approved?

**Comment by consideRatio** [comment]
I don't want to be the sole approver of adding that package but I'm positively inclined for that change as I see it doesn't seem to depend on lots of other python libraries etc in it.

You can go ahead and open a PR about that. It should be added to requirements.in and then the images/hub/dependencies script should run as well, hmm, which will update lots of dependencies I think - but yes that is what should be done I think.

---

@weisdd what does the `-e` part following `/usr/bin/bash` do? I don't see it listed in `man bash` output.

**Comment by weisdd** [comment]
@consideRatio `-e` is documented here: https://www.gnu.org/software/bash/manual/bash.html#Modifying-Shell-Behavior
Basically, it makes the whole script fail if any of the commands within the pipeline exit with non-zero status. For example:

**with -e**:

```bash
cat << EOF > 2.sh
#!/usr/bin/bash
set -exo pipefail
sdfsdfsdf
echo "test"
EOF
chmod +x 2.sh
```

```shell
./2.sh 
+ sdfsdfsdf
./2.sh: line 3: sdfsdfsdf: command not found
```

**without -e**:

```bash
cat << EOF > 2.sh
#!/usr/bin/bash
set -xo pipefail
sdfsdfsdf
echo "test"
EOF
chmod +x 2.sh
```

```shell
./2.sh
+ sdfsdfsdf
./2.sh: line 3: sdfsdfsdf: command not found
+ echo test
test
```

As you can see, in the second example, the script continues to run despite the failure along the way (missing "sdfsdfsdf" binary).

So, basically, I want jupyterhub container to fail to start in case pip cannot install the DB driver. Of course, jupyterhub will eventually fail anyway in that case (sqlalchemy verifies urls against the list of installed drivers), so `-e` is redundant. It's more like a habbit to add it everywhere to make scripts fail early :) I even added it here twice :D (in `-ec` and in `set -exo`) - As long as options are set to the same values, it's fine :)
